---
title: "Import cancer dataset from TCGA"
author: "André Veríssimo"
output:
  github_document:
    toc: yes
  rmarkdown::html_vignette:
    toc: yes
vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
params:
  project: "TCGA-PRAD"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning=FALSE)
library(futile.logger)
```

# Package information

## How to use the dataset

1. Install `r params$project`.data by using `devtools` package.

1. Load the library

1. Load the required datasets (one or more of the following)
    - `multiAssay`
    - `gdc.original`

### Example *(for BRCA)*

```{r install_brca.data, eval=FALSE}
# The library can also be loaded and use the function install_git without 'devtools::' prefix
BiocManager::install('https://github.com/averissimo/tcga.data/releases/download/2016.12.15-brca/brca.data_1.0.tar.gz')
#
# Load the brca.data package
library(brca.data)
# start using the data
data(multiAssay)
multiAssay
```

## Description of the data set

The Cancer Genome Atlas `r params$project` data collection is part of a larger effort to build a research community focused on connecting cancer phenotypes to genotypes by providing clinical images matched to subjects from The Cancer Genome Atlas ([TCGA](https://cancergenome.nih.gov/)). 

The data is publicly available (https://gdc-portal.nci.nih.gov/) and is decomposed into two data sets:

1. the gene expression data;

1. the clinical data.

### Explaining the TCGA codes

The following links explain (1) the individuals' barcode and (2) the sample type code:

1.  [Link to individuals' barcode from tcga](https://wiki.nci.nih.gov/display/TCGA/TCGA+barcode?desktop=true&macroName=unmigrated-inline-wiki-markup)

1. [Link to sample type code from tcga](https://gdc.cancer.gov/resources-tcga-users/tcga-code-tables/sample-type-codes)

Explanation of TCGA barcode using example `TCGA-02-0001-01C-01D-0182-01`

![Details on TCGA Barcode](https://wiki.nci.nih.gov/download/attachments/39294833/barcode.png?version=1&modificationDate=1300400318000&api=v2)

| Label       | Identifier for     | Value | Value Description | Possible Values |
|-------------|--------------------|-------|-------------------|-----------------|
| Project     | Project name       | TCGA | TCGA project | TCGA |
| TSS         | Tissue source site | 02 | GBM (brain tumor) sample from MD Anderson | See [Code Tables Report](https://gdc.cancer.gov/resources-tcga-users/tcga-code-tables/tissue-source-site-codes) |
| Participant | Study participant | 0001 | The first participant from MD Anderson for GBM study | Any alpha-numeric value |
| Sample      | Sample type       | 01 | A solid tumor | Tumor types range from 01 - 09, normal types from 10 - 19 and control samples from 20 - 29. See [Code Tables Report](https://gdc.cancer.gov/resources-tcga-users/tcga-code-tables/sample-type-codes) for a complete list of sample codes |
| Vial        | Order of sample in a sequence of samples       | C | The third vial | A to Z |
| Portion     | Order of portion in a sequence of 100 - 120 mg sample portions | 01 | The first portion of the sample | 01-99 |
| Analyte     | Molecular type of analyte for analysis         | D | The analyte is a DNA sample | See [Code Tables Report](https://gdc.cancer.gov/resources-tcga-users/tcga-code-tables/portion-analyte-codes) |
| Plate       | Order of plate in a sequence of 96-well plates | 0182 | The 182nd plate | 4-digit alphanumeric value |
| Center      | Sequencing or characterization center that will receive the aliquot for analysis | 01 | The Broad Institute GCC | See [Code Tables Report](https://gdc.cancer.gov/resources-tcga-users/tcga-code-tables/center-codes) |

# Script to generate the package data

```{r}
#
#
#
#
# This now documents the generation of the dataset.
#
# No longer a generic documentation of the package!
#
# 
#
```

## Loading data from TCGA

### TCGAbiolinks package

The data set can also be extracted through the [TCGAbiolinks R bioconductor package](https://bioconductor.org/packages/release/bioc/html/TCGAbiolinks.html)

```{r install_biolinks, eval=FALSE}
BiocManager::install("TCGAbiolinks")
```

#### Download data (project: *`r params$project`*)

Change `project` variable to get other data projects in TCGA.

```{r biolinks}
library(TCGAbiolinks)
project <- params$project
cached.file <- file.path('..', sprintf('tcga-biolinks-%s-cache.RData', tolower(gsub('TCGA-', '', params$project))))
if (file.exists(cached.file)) {
  load(cached.file)
} else {
  query <- list()
  # Query the GDC to obtain data
  flog.info('Preparing query for clinical data...')
  query$clinical <- GDCquery(project       = project,
                             data.category = "Clinical",
                             file.type     = "xml")
  # Download query data
  flog.info('Downloading clinical data...')
  GDCdownload(query$clinical, method = 'api')

  flog.info('Organizing clinical data...')
  # Prepare R object with data
  gdc <- list()
  gdc$clinical        <- GDCprepare_clinic(query$clinical, clinical.info = 'patient')
  gdc$drug            <- GDCprepare_clinic(query$clinical, clinical.info =  'drug')
  gdc$radiation       <- GDCprepare_clinic(query$clinical, clinical.info =  'radiation')
  gdc$admin           <- GDCprepare_clinic(query$clinical, clinical.info =  'admin')
  gdc$follow.up       <- GDCprepare_clinic(query$clinical, clinical.info =  'follow_up')
  gdc$stage_event     <- GDCprepare_clinic(query$clinical, clinical.info =  'stage_event')
  gdc$new_tumor_event <- GDCprepare_clinic(query$clinical, clinical.info =  'new_tumor_event')
  #
  flog.info('Preparing query for biospecimen data...')
  query$biospecimen <- GDCquery(project       = project,
                                data.category = "Biospecimen",
                                data.type     = 'Biospecimen Supplement',
                                file.type     = "xml")

  #
  flog.info('Download biospecimen data...')
  GDCdownload(query$biospecimen, method = 'api')
  #
  flog.info('Organizing biospecimen data...')
  gdc$bio.protocol  <- 'There was a problem downloading from GDC, not available'
  gdc$bio.admin     <- 'There was a problem downloading from GDC, not available'
  gdc$bio.aliquot   <- 'There was a problem downloading from GDC, not available'
  gdc$bio.analyte   <- 'There was a problem downloading from GDC, not available'
  gdc$bio.patient   <- 'There was a problem downloading from GDC, not available'
  gdc$bio.sample    <- 'There was a problem downloading from GDC, not available'
  gdc$bio.portion   <- 'There was a problem downloading from GDC, not available'
  gdc$bio.slide     <- 'There was a problem downloading from GDC, not available'
  tryCatch({
    gdc$bio.protocol  <- GDCprepare_clinic(query$biospecimen, clinical.info = 'protocol')
    gdc$bio.admin     <- GDCprepare_clinic(query$biospecimen, clinical.info = 'admin')
    gdc$bio.aliquot   <- GDCprepare_clinic(query$biospecimen, clinical.info = 'aliquot')
    gdc$bio.analyte   <- GDCprepare_clinic(query$biospecimen, clinical.info = 'analyte')
    gdc$bio.patient   <- GDCprepare_clinic(query$biospecimen, clinical.info = 'bio_patient')
    gdc$bio.sample    <- GDCprepare_clinic(query$biospecimen, clinical.info = 'sample')
    gdc$bio.portion   <- GDCprepare_clinic(query$biospecimen, clinical.info = 'portion')
    gdc$bio.slide     <- GDCprepare_clinic(query$biospecimen, clinical.info = 'slide')
    #
  
    # only pre-processing needed for bio.sample
    rownames(gdc$bio.sample) <- gdc$bio.sample$bcr_sample_barcode
  }, error = function(err) {
    flog.warn('Error preparing biosample data, %s', err)
  })
  
  flog.info('Preparing query for RNASeq FPKM data...')
  query$rnaseq.fpkm <- GDCquery(project = project,
                           data.category = "Transcriptome Profiling",
                           data.type     = "Gene Expression Quantification",
                           workflow.type = "HTSeq - FPKM") 
  #

  flog.info('Downloading RNASeq FPKM data...')
  GDCdownload(query$rnaseq.fpkm, method = 'api', files.per.chunk = 6)
  
  flog.info('Preparing query for RNASeq Counts data...')
  query$rnaseq.counts <- GDCquery(project = project,
                           data.category = "Transcriptome Profiling",
                           data.type     = "Gene Expression Quantification",
                           workflow.type = "HTSeq - Counts")
  
  #
  flog.info('Downloading RNASeq Counts data...')
  GDCdownload(query$rnaseq.counts, method = 'api', files.per.chunk = 6)
  
  flog.info('Organizing RNASeq data...')
  gdc$rnaseq.fpkm   <- GDCprepare(query$rnaseq.fpkm)
  gdc$rnaseq.counts <- GDCprepare(query$rnaseq.counts)
  #
  flog.info('Preparing query for mutation data...')
  query$mutation <- GDCquery(project = project,
                             data.category = "Simple Nucleotide Variation",
                             data.type     = "Masked Somatic Mutation",
                             workflow.type = "MuSE Variant Aggregation and Masking")
  flog.info('Downloading mutation data...')
  GDCdownload(query$mutation, method = 'api', files.per.chunk = 6)
  
  flog.info('Organizing mutation data')

  gdc$mutation <- GDCprepare(query$mutation)
  # clean up table by removing all rows that are NA
  gdc$mutation <- gdc$mutation[,colSums(is.na(gdc$mutation)) != nrow(gdc$mutation)]
  #
  save(gdc, query,
       file = cached.file)
}
```

### Auxiliary libraries and functins to be used

```{r load_libraries}
# Library to print some information
library(futile.logger)
library(ggplot2)
library(tibble)
library(dplyr)
library(Matrix)

devtools::load_all()
```

### Extracting all the barcodes

Pattern to retrieve TCGA patient and sample type from extended barcode.

```{r all_data_set}
#
# clinical data
rownames(gdc$clinical) <- gdc$clinical$bcr_patient_barcode
gdc$clinical           <- gdc$clinical
clinical               <- gdc$clinical
follow.up              <- gdc$follow.up
```

## Processing the data

### Store patient's clinical data

```{r clinical}
#
#
levels(clinical$vital_status) <- levels(clinical$vital_status) %>% {replace(., . == '', NA)}
levels(follow.up$vital_status)    <- levels(follow.up$vital_status) %>% {replace(., . == '', NA)}

updated.surv           <- update.survival.from.followup(clinical, follow.up) %>% as.data.frame
rownames(updated.surv) <- updated.surv$bcr_patient_barcode

clinical$surv_event_time <- updated.surv[clinical$bcr_patient_barcode, 'surv_event_time']
clinical$vital_status    <- updated.surv[clinical$bcr_patient_barcode, 'vital_status']
```

### Mutations

```{r discard_dups_in_same_position}
mutation <- list()
# remove duplicate mutations on same position
mutation$data <- distinct(gdc$mutation, Hugo_Symbol, Chromosome, Start_Position, End_Position, Strand, .keep_all = T)
dups.ix       <- paste0(gdc$mutation$Hugo_Symbol, gdc$mutation$Chromosome, gdc$mutation$Start_Position, gdc$mutation$End_Position, gdc$mutation$Strand)
#
# check if the number of duplicated are the same
# flog.info('Is the number of discarded row from distict the same as duplicated? R: %s', nrow(gdc$mutation) - nrow(mutation) == sum(duplicated(dups.ix)))
#
#
mutation$dups <- gdc$mutation[duplicated(dups.ix) | duplicated(dups.ix, fromLast = TRUE),]
mutation$dups <- dplyr::arrange(mutation$dups, Hugo_Symbol, Tumor_Sample_Barcode)
```

#### Recalculate Mutations count of (row: gene) x (col: case)

```{r calc_mutations_count}
mutation$count <- matrix(integer(1), ncol = nrow(gdc$clinical), nrow = length(unique(mutation$data$Gene)), 
                         dimnames = list(unique(mutation$data$Gene), gdc$clinical$bcr_patient_barcode))
# get a temporary index list for fast access
ix.list <- cbind(mutation$data$Gene, getParticipantCode(mutation$data$Tumor_Sample_Barcode))

ix.list <- ix.list[!is.na(mutation$data$Gene),]
ix.list <- as.tibble(ix.list, stringsAsFactors = FALSE)
colnames(ix.list) <- c('gene', 'patient')
# this is a long run, should be as simple as possible
sum.list <- ix.list %>% group_by(gene, patient) %>% summarise(n = n())
for(ix in seq(nrow(sum.list))) {
  mutation$count[sum.list[[ix,'gene']], sum.list[[ix,'patient']]] <- sum.list[[ix,'n']]
}
mutation$count <- Matrix::Matrix(mutation$count, sparse = TRUE)
```

```{r mutation_hist}
# analyse raw mutation results
mutations.by.case <- data.frame(ix = seq(ncol(mutation$count)), case = colSums(mutation$count), unique = colSums(mutation$count != 0))
mutations.by.case$rel.unique.case <- mutations.by.case$unique / mutations.by.case$case
#
flog.info('Summary of count of mutations (with repeated)', summary(mutations.by.case$case), capture = TRUE)
flog.info('Summary of count of mutations (unique only)', summary(mutations.by.case$unique), capture = TRUE)
#
ggplot(data = mutations.by.case) + 
  theme_minimal() +
  geom_point(data = base::subset(mutations.by.case, rel.unique.case <= 1.2), aes(x = ix, y = unique, colour = rel.unique.case)) + 
  geom_point(data = base::subset(mutations.by.case, rel.unique.case > 1.2),  aes(x = ix, y = unique, colour = rel.unique.case)) + 
  geom_point(data = base::subset(mutations.by.case, rel.unique.case > 1.4),  aes(x = ix, y = unique, colour = rel.unique.case)) + 
  geom_point(data = base::subset(mutations.by.case, rel.unique.case > 1.6),  aes(x = ix, y = unique, colour = rel.unique.case)) + 
  geom_point(data = base::subset(mutations.by.case, rel.unique.case > 1.8),  aes(x = ix, y = unique, colour = rel.unique.case)) + 
  scale_colour_gradient('% of unique mutations per gene', high = 'red', low = 'green') +
  labs(title="Histogram for Mutations by Case") +
  labs(x='Individual', y="Count of (unique) mutations") +
  theme(legend.position = 'top') +
  scale_y_continuous(breaks = c(1, 10, 100, 1000, 10000), trans = 'log10')
```

## MultiAssayExperiment

Builds a multiAssayExperiment with clinical, rnaseq and mutation count data

```{r}
multiAssay <- build.assay(clinical.custom      = clinical, 
                          gdc.custom           = gdc, 
                          mutation.custom      = mutation$count)
multiAssay
```

## Exported data

- `gdc.original`: original data taken from TCGABiolinks, with the exception of expression data;
- `multiAssay`: MultiExperimentAssay object containing clinical, RNASeq FPKM, RNASeq Counts and mutation count;

```{r export_data, include=FALSE, eval=TRUE}
# Extract all expression data to a different variable to remove redundancy in `tissue`
gdc.original <- gdc
gdc.original$rnaseq.counts <- NULL
gdc.original$rnaseq.fpkm   <- NULL
devtools::use_data(gdc.original, overwrite = TRUE)
devtools::use_data(multiAssay,   overwrite = TRUE)
```

```{r, include=FALSE, eval=FALSE}
rmarkdown::render('build_data.Rmd')
```
